APE25.rst

```diff
@@ -0,0 +1,166 @@
+An initial roadmap for static type checking
+-------------------------------------------
+
+author: Nick Murphy
+
+date-created:
+
+date-last-revised:
+
+date-accepted:
+
+type: Standard Track
+
+status: Discussion
+
+
+Abstract
+--------
+
+This APE describes the initial process for gradually adding static
+type checking to Astropy.
+
+Detailed description
+--------------------
+
+Type hint annotations allow us to specify the expected data types of
+function arguments, return values, and variable assignments. For
+example,
+
+.. code-block::
+
+   def f(x: int) -> float:
+       ...
+
+indicates that ``x`` should be an ``int`` and the return value should
+be a ``float``. Type hints are not enforced by Python during
+runtime. Type hints can be used by static type checkers to find
+problems with the code and by Jupyter notebooks and integrated
+development environments (IDEs) to help with code completion. Type
+hints also serve as a form of documentation in function and method
+signatures.
+
+For most of its history, type hint annotations have not been applied
+within Astropy's source code (though annotations have been used to
+indicate the expected units or physical types of function arguments
+and return values). Only recently have type hint annotations begun to
+be applied to Astropy. However, no mechanism is in place to check the
+correctness and validity of type hint annotations. ...
+
+Most guides on adding type hint annotations and enabling static type
+checking recommend starting small, running mypy consistently,
+prioritize annotating widely used imports, add type hint annotations
+for new code, gradually increase the fraction of the code base that is
+checked my mypy, and gradually increase the strictness of the checks
+performed by mypy.
+
+.. This section describes the need for the APE.  It should describe
+   the existing problem that it is trying to solve and why this APE
+   makes the situation better.  It should include examples of how the
+   new functionality would be used and perhaps some use cases.
+
+
+Benefits of static type checking
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+#. Can find programming errors that would be difficult to find
+   otherwise
+#. Helps introspection and tab completion in Jupyter notebooks
+#. Helps IDEs to make things easier for humans
+#. Helpful for downstream package maintainers
+
+Disadvantages of static type checking
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+#. Requiring type 
+
+1. Requiring type hint annotations that mypy doesn't complain about
+   adds an extra step to contributing to Astropy...
+
+1. An additional step is required to contribute to Astropy.
+
+1. Type hint annotations may add a barrier to entry for contributing
```

Comment:
This also *feels* similar to the first two points, but I guess maybe you mean to say that existing type hints may make the code harder to read from a newcomer's perspective ? This is something I've heard in similar contexts, coming from _seasoned_ devs, and I'm honestly not convinced it's a valid point, but it's nonetheless an opinion that exists.